flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/writer/topic/TopicProducerRegister.java

```diff
@@ -189,4 +198,16 @@ private void clearTransactions() {
         // Clear the transactions, we would create new transaction when new message comes.
         transactionRegister.clear();
     }
+
+    /** Retrieve the producer metrics and map to Flink SinkWriterMetrics. */
+    private void updateProducerStats() {
+        long numBytesOutSum = 0;
+        long numRecordsOutSum = 0;
+        for (Producer<?> producer : producerRegister.values()) {
+            numBytesOutSum += producer.getStats().getNumBytesSent();
+            numRecordsOutSum += producer.getStats().getNumMsgsSent();
```

Comment:
recordsOut can be collected on sendAsync() callback complete, but I think current implementation is ok ~